File: algoritmo/busquedaBinaria.js
/**
 * Realiza una búsqueda binaria en un arreglo ordenado.
 * 
 * Precondiciones (Contratos):
 * - El primer parámetro debe ser un Arreglo (Array).
 * - El objetivo debe ser un número entero.
 * - El arreglo debe estar ordenado de menor a mayor.
 * 
 * Postcondiciones (Contratos):
 * - Retorna el índice correcto si el elemento existe (lista[indice] === objetivo).
 * - Retorna -1 si el elemento no existe en la lista.
 */
function busquedaBinaria(lista, objetivo) {
    // 1. Verificación de Tipos (Precondiciones)
    if (!Array.isArray(lista)) {
        throw new TypeError("El primer parámetro debe ser un arreglo.");
    }
    if (!Number.isInteger(objetivo)) {
        throw new TypeError("El objetivo debe ser un número entero.");
    }

    // 2. Verificación de ordenamiento (Precondición de Contrato)
    const esOrdenada = lista.every((val, i) => i === 0 || lista[i - 1] <= val);
    if (!esOrdenada) {
        throw new Error("El arreglo de entrada debe estar ordenado.");
    }

    let izquierda = 0;
    let derecha = lista.length - 1;
    
    // Salvaguarda contra bucles infinitos (vital para mutation testing)
    const limiteIteraciones = lista.length + 5;
    let iteraciones = 0;

    while (izquierda <= derecha) {
        iteraciones++;
        if (iteraciones > limiteIteraciones) {
            throw new Error("Límite de iteraciones excedido: bucle infinito detectado.");
        }

        const medio = Math.floor((izquierda + derecha) / 2);
        const valorMedio = lista[medio];

        if (valorMedio === objetivo) {
            // Postcondición: verificar correspondencia de valor
            if (lista[medio] !== objetivo) {
                throw new Error("Fallo de postcondición: el valor medio no coincide.");
            }
            return medio;
        } else if (valorMedio < objetivo) {
            izquierda = medio + 1;
        } else {
            derecha = medio - 1;
        }
    }

    // Postcondición: verificar que el elemento no está en el arreglo
    if (lista.includes(objetivo)) {
        throw new Error("Fallo de postcondición: el elemento sí existía en la lista.");
    }
    return -1;
}

module.exports = { busquedaBinaria };

File: algoritmo/busqueda_binaria.py
def busqueda_binaria(lista: list, objetivo: int) -> int:
    """
    Realiza una busqueda binaria en una lista ordenada.
    
    Precondiciones (Contratos):
    - La lista debe ser de tipo list.
    - El objetivo debe ser de tipo int.
    - La lista debe estar ordenada de menor a mayor.
    
    Postcondiciones (Contratos):
    - Retorna el indice correcto si el elemento existe en la lista (lista[indice] == objetivo).
    - Retorna -1 si el elemento no existe en la lista.
    """
    # 1. Verificacion de tipos (Precondiciones)
    if not isinstance(lista, list):
        raise TypeError("El primer parametro debe ser una lista.")
    if not isinstance(objetivo, int):
        raise TypeError("El objetivo debe ser un entero.")
    
    # 2. Verificacion de ordenamiento (Precondicion de Contrato)
    es_ordenada = all(lista[i] <= lista[i + 1] for i in range(len(lista) - 1))
    if not es_ordenada:
        raise ValueError("La lista de entrada debe estar ordenada.")
        
    izquierda = 0
    derecha = len(lista) - 1
    
    # Salvaguarda contra bucles infinitos (vital para mutation testing)
    limite_iteraciones = len(lista) + 5
    iteraciones = 0

    while izquierda <= derecha:
        iteraciones += 1
        if iteraciones > limite_iteraciones:
            raise RuntimeError("Limite de iteraciones excedido: bucle infinito detectado.")
            
        medio = (izquierda + derecha) // 2
        valor_medio = lista[medio]

        if valor_medio == objetivo:
            # Postcondicion: verificar que el elemento en el indice retornado es el buscado
            assert lista[medio] == objetivo, "Fallo de postcondicion: el valor medio no coincide"
            return medio
        elif valor_medio < objetivo:
            izquierda = medio + 1
        else:
            derecha = medio - 1

    # Postcondicion: verificar que el elemento realmente no esta en la lista
    assert objetivo not in lista, "Fallo de postcondicion: el elemento si existia en la lista"
    return -1
